File: frontend-v2/src/api/analytics.ts
import { apiClient } from './client';
import type {
  FunnelResponse,
  DecisionBreakdownResponse,
  ThroughputResponse,
  TimeInStageResponse,
  JobVolumeResponse,
} from '../types';

export const analyticsApi = {
  getFunnel: (jobId?: number): Promise<FunnelResponse> =>
    apiClient.get('/analytics/funnel', { params: jobId != null ? { job_id: jobId } : {} }),

  getDecisionBreakdown: (jobId?: number): Promise<DecisionBreakdownResponse> =>
    apiClient.get('/analytics/decisions', { params: jobId != null ? { job_id: jobId } : {} }),

  getThroughput: (days = 30, jobId?: number): Promise<ThroughputResponse> =>
    apiClient.get('/analytics/throughput', { params: { days, ...(jobId != null ? { job_id: jobId } : {}) } }),

  getTimeInStage: (jobId?: number): Promise<TimeInStageResponse> =>
    apiClient.get('/analytics/time-in-stage', { params: jobId != null ? { job_id: jobId } : {} }),

  getJobVolume: (): Promise<JobVolumeResponse> => apiClient.get('/analytics/job-volume'),
};

File: app/api/analytics.py
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.schemas.analytics import (
    FunnelResponse,
    DecisionBreakdownResponse,
    ThroughputResponse,
    TimeInStageResponse,
    JobVolumeResponse,
)
from app.services.analytics import analytics_service

router = APIRouter()


@router.get("/funnel", response_model=FunnelResponse)
async def get_funnel(job_id: Optional[int] = Query(None), db: AsyncSession = Depends(get_db)):
    return await analytics_service.funnel(db, job_id=job_id)


@router.get("/decisions", response_model=DecisionBreakdownResponse)
async def get_decisions(job_id: Optional[int] = Query(None), db: AsyncSession = Depends(get_db)):
    return await analytics_service.decision_breakdown(db, job_id=job_id)


@router.get("/throughput", response_model=ThroughputResponse)
async def get_throughput(
    days: int = Query(30, ge=1, le=365),
    job_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    return await analytics_service.throughput(db, days=days, job_id=job_id)


@router.get("/time-in-stage", response_model=TimeInStageResponse)
async def get_time_in_stage(job_id: Optional[int] = Query(None), db: AsyncSession = Depends(get_db)):
    return await analytics_service.time_in_stage(db, job_id=job_id)


@router.get("/job-volume", response_model=JobVolumeResponse)
async def get_job_volume(db: AsyncSession = Depends(get_db)):
    return await analytics_service.job_volume(db)
